### PR TITLE
Propagator

### PR DESCRIPTION
**Pull request by jaescartin1**
Decoupled timestep with a Propagator class common to all the tests.

This is the first step, generalize the loop. 

The second step, not implemented on this branch, will be add a list of conservedQuantities , maybe different between tests.

As Sebastian suggested, something like:

ParticlesData d = SedovDataGenerator...
// x, y, z, h, m are mandatory and assumed conserved
d.addConserved(&d.mui, &d.u, &d.vx, &d.vy, &d.vz, ...);
d.addSecondary(&d.ro, &d.p, &d.c, ...);

for (d.iteration = 0; d.iteration < maxStep; d.iteration++)
{
    propagator.hydroStep(domain, d);
    // file output if needed
}

but the second step (conserved and secondary quantities) should be develop in another PR

### CODE REVIEW

File: include/tests/propagator.hpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    void step(const std::string &name){
+         timer.step(name);
+    }
+
+    size_t neighbors(){
+        return neighborsSum(taskList.tasks);
+    }
+
+    // Advance simulation by one step
+    void hydroStep()
```
Comment (sekelle) [inline include/tests/propagator.hpp]: Take the domain and ParticlesData here as arguments.

File: include/tests/propagator.hpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+#include "cstone/domain/domain.hpp"
+
+#include "sph/timestep.hpp"
+
+using namespace cstone;
+using namespace sphexa;
+using namespace sphexa::sph;
+
+template <class I, class T, class Accelerator = CpuTag>
```
Comment (sekelle) [inline include/tests/propagator.hpp]: These template arguments go to `hydroStep`, but as
`template<class DomainType, class ParticleDataType>`

File: include/tests/propagator.hpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    MasterProcessTimer         timer;                           //
+
+    const size_t               nTasks;                          //
+    const size_t               ngmax;                           //
+    const size_t               ng0;                             //
+
+    TaskList                   taskList;                        //
+
+    const bool                 gravity;                         //
```
Comment (sekelle) [inline include/tests/propagator.hpp]: You can remove anything related to gravity, we're going to make a separate class Propagator for gravity.
(There will be a lot of changes to how gravity is handled)

File: include/tests/propagator.hpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+ * @author Jose A. Escartin <ja.escartin@gmail.com>
+ */
+
+#pragma once
+
+#include <iostream>
+
+#include "../Timer.hpp"
+#include "../Task.hpp"
+#include "../ParticlesData.hpp"
```
Comment (sekelle) [inline include/tests/propagator.hpp]: remember to remove "../" once this file is in `include` (you can actually move it already now)

File: include/tests/propagator.hpp [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        timer.stop();
+    }
+
+    void step(const std::string &name){
+         timer.step(name);
+    }
+
+    size_t neighbors(){
+        return neighborsSum(taskList.tasks);
+    }
```
Comment (sekelle) [inline include/tests/propagator.hpp]: The back-and-force between timing inside the propagator and outside it adds a lot of clutter.
Do you think you could move all the timing inside the propagator and not time the file I/O, such that
the file I/O only is part of the final runtime, but not the iteration steps?

**Comment by sekelle** [comment]
Note on formatting: some changes are not compatible with the .clang-format in the top-level directory.
I'll address that if you don't have a formatting-tool at hand.

**Comment by jaescartin1** [comment]
Hi Sebastian,
       all the changes that you required were applied:
- hydrostep with domain and particleData as parameters
- template in function hydrostep, instead of in the class
- Removed gravity boolean and made another function in propagator with gravity
- Moved include/tests/propagator.hpp to include/ and renamed the includes inside
- Removed timer from propagator and included in app codes
- Now Public access to the parameter taskList, in order to reduce the access overhead

About formating, please, let me know the tool that give the format that you wish.

File: include/propagator.hpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        , taskList(0, nParticles, nTasks, ngmax, ng0)
+    {}
+
+    size_t neighbors(){
+        return neighborsSum(taskList.tasks);
+    }
+
+    template <class DomainType, class ParticleDataType, class T>
+    void hydroStep(DomainType& domain, ParticleDataType& d)
+    {
```
Comment (sekelle) [inline include/propagator.hpp:83]:
```suggestion
    {
    using T = ParticleDataType::RealType;
```

File: include/propagator.hpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        , ngmax(ngmax)
+        , ng0(ng0)
+        , taskList(0, nParticles, nTasks, ngmax, ng0)
+    {}
+
+    size_t neighbors(){
+        return neighborsSum(taskList.tasks);
+    }
+
+    template <class DomainType, class ParticleDataType, class T>
```
Comment (sekelle) [inline include/propagator.hpp]:
```suggestion
    template <class DomainType, class ParticleDataType>
```

File: include/propagator.hpp [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+private:
+
+    const size_t       nTasks;                                  //
+    const size_t       ngmax;                                   //
+    const size_t       ng0;                                     //
+
+public:
+
+    TaskList           taskList;                                //
+    MasterProcessTimer timer;                                   //
```
Comment (sekelle) [inline include/propagator.hpp:61]: can be private?

File: src/evrard/evrard.cpp [hunk 9]
```diff
@@ -101,126 +105,120 @@ int main(int argc, char** argv)
...
-        computePositions<Real, computeAcceleration<Real, Dataset>>(taskList.tasks, d, domain.box());
-        timer.step("UpdateQuantities");
-        computeTotalEnergy<Real>(taskList.tasks, d);
-        d.etot += d.egrav;
-        timer.step("EnergyConservation"); // AllReduce(sum:ecin,ein)
-        updateSmoothingLength<Real>(taskList.tasks, d);
-        timer.step("UpdateSmoothingLength");
-
-        size_t totalNeighbors = neighborsSum(taskList.tasks);
+        propagator.hydroStepGravity<DomainType, Dataset, Real>(domain, d);
```
Comment (sekelle) [inline src/evrard/evrard.cpp]:
```suggestion
        propagator.hydroStepGravity(domain, d);
```
Should be possible once the last template parameter is removed

**Review by sekelle — APPROVED** [review]

**Comment by sekelle** [comment]
cscs-ci run

**Pull request merged**